### PR TITLE
feat: Implementation of Fire2.0 and ABC Fire

### PR DESCRIPTION
**Pull request by sebastianecheverrir**
### Implementation of version 2.0 of the optimiser FIRE, including support for ABC-FIRE. The algorithms were taken from: 

        * Guénolé et al., Comput. Mater. Sci. 175, 109584 (2020).
        * Echeverri Restrepo & Andric, Comput. Mater. Sci. 218, 111978 (2023). 

### Changes:

Fire 2.0 and ABC-FIRE

        * src/kups/relaxation/optax/__init__.py
        * src/kups/relaxation/optax/fire.py
        * src/kups/relaxation/optax/optimizer.py

New example added to test the performance of ABC-FIRE on a system containing a Fe BCC crystal with vacancies

        * examples/host/Fe_bcc_vac_perturbed.cif
        * examples/relax_mace_fe_vac_fire2_abc.yaml

Added additional output to tqdm to show the energy and the max force during minimisation ( E[eV],  fmax[eV/Å])

        * src/kups/application/relaxation/simulation.py

### CODE REVIEW

**Comment by sebastianecheverrir** [comment]
GENAI was used to assist with the coding of the current pull request

**Review by jonkhler — COMMENTED** [review]
Thanks for upstreaming this. Having FIRE 2.0 + ABC-FIRE in kUPS is genuinely useful, and it's nice to see the algorithm contributed by one of the ABC-FIRE authors. The implementation reads as a faithful port of LAMMPS' `min_fire.cpp` `eulerimplicit` integrator, and the documented deviation (no force recompute inside the half-step backtrack) is the right call for a single-pass Optax transform.

A few things I'd like to see addressed before merging:

**Required for merge**

1. **Tests for `scale_by_fire2`.** The function has four orthogonal flags (`use_abc`, `halfstepback`, `delaystep_start`, `max_step is None`) plus the P>0 / P≤0 branch, and none of them are covered. Could you mirror the structure of `test_fire.py` and add coverage for: init state, dt adaptation in both branches, the half-step backtrack on P≤0, the `delaystep_start` startup mask, ABC bias correction at small N, and one end-to-end convergence test on a quadratic? Doesn't need to be exhaustive, just enough that a future refactor can't silently break a branch.

2. **Split the tqdm postfix change into its own PR.** The energy + fmax display in `simulation.py` is a nice UX improvement but unrelated to FIRE. Easier to land as a small standalone PR; also keeps this one focused on the optimizer.

3. **Document the `max_step` semantics divergence.** `scale_by_fire` clips position updates by global L2 norm (default 0.2 Å); `scale_by_fire2` uses LAMMPS-style ∞-norm `dmax` (default 0.1 Å, with a different mechanism for the ABC variant). A user swapping one for the other in their YAML gets quietly different bounds. A note in the `scale_by_fire2` docstring contrasting the two would prevent confusion.

**Nice-to-have**

- A short docstring pointer to the paper figures (or a one-line empirical note) showing where FIRE 2.0 / ABC-FIRE wins over FIRE 1.0 would help users pick between them.

**On our plate, not yours.** There's a related concern around per-system optimizer state for batched relaxation of independent systems (see #95) that will affect `scale_by_fire2` the same way it affects L-BFGS: shared scalar `dt` / `alpha` / `n_pos` across systems isn't ideal. We'll handle that when we land the per-system state pattern; nothing for you to change here.

Once 1 to 3 are in I'll do another pass and merge. Happy to help if any of this is unclear, and if you'd prefer I cherry-pick the tqdm change into a separate PR with you as Co-authored-by, just say the word.

File: src/kups/relaxation/optax/fire.py [hunk 1]
```diff
@@ -143,3 +170,292 @@ def update_fn(
...
+            when velocity is zero.
+
+    Returns:
+        Optax GradientTransformation implementing FIRE 2.0.
+
+    References:
+        * Guénolé et al., Comput. Mater. Sci. 175, 109584 (2020).
+        * Echeverri Restrepo & Andric, Comput. Mater. Sci. 218, 111978 (2023).
+        * LAMMPS ``src/min_fire.cpp`` (develop branch).
+    """
```
Comment (jonkhler) [inline src/kups/relaxation/optax/fire.py:331]: The `max_step` description above is great for understanding FIRE 2.0's behavior in isolation, but worth contrasting explicitly with `scale_by_fire`'s semantics so users don't switch between them and silently get different step bounds. Something like adding to the bottom of the docstring:

```
Note:
    Differs from ``scale_by_fire``, which clips position updates by
    global L2 norm (default 0.2 Å). Switching from ``scale_by_fire``
    to ``scale_by_fire2`` will produce different step bounds even
    with the same ``max_step`` value.
```

File: src/kups/application/relaxation/simulation.py [hunk 2]
```diff
@@ -152,8 +152,13 @@ def converged(s: State) -> bool:
...
         return bool(max_force < config.force_tolerance)
 
+    def _postfix(s: State) -> dict[str, Any]:
+        e = jnp.asarray(s.systems.data.potential_energy).sum()
+        fmax = jnp.max(jnp.linalg.norm(s.particles.data.forces, axis=-1))
+        return {"E[eV]": f"{float(e): .6f}", "fmax[eV/Å]": f"{float(fmax): .4e}"}
+
     logger = CompositeLogger(
-        TqdmLogger(config.max_steps),
+        TqdmLogger(config.max_steps, postfix=_postfix),
```
Comment (jonkhler) [inline src/kups/application/relaxation/simulation.py:161]: Could you pull this energy + fmax tqdm postfix change into a separate small PR? It's a nice UX improvement and I'd like to land it, but it's unrelated to FIRE 2.0. Easier to merge it standalone, and keeps this PR focused on the optimizer. Happy to fast-track the split-out PR (or cherry-pick with you as Co-authored-by if that's easier on your side).

**Comment by n-gao** [comment]
Thanks for the great contribution @sebastianecheverrir, I agree with Jonas' comments, it would be gerat if we could isolate changes here on focus on the algorithm itself. It would also be super helpful if you could add some unit tests!

Though I think we can land this and then adapt to fix #94.

Thanks a lot :)

**Comment by sebastianecheverrir** [comment]
I think all the comments regarding the optimiser have been addressed in the last commit. About the tqdm fix, please fell free to cherry pick it.

File: examples/host/Fe_bcc_vac_perturbed.cif
Comment (n-gao) [inline examples/host/Fe_bcc_vac_perturbed.cif:1]: Please remove.

File: examples/relax_mace_fe_vac_fire2_abc.yaml
Comment (n-gao) [inline examples/relax_mace_fe_vac_fire2_abc.yaml:1]: Please remove.

**Review by n-gao — CHANGES_REQUESTED** [review]
Please remove the unrelated files to this PR and we are good to go :)

**Comment by sebastianecheverrir** [comment]
The files have now been removed. Thx for your quick action with this pull request!

**Comment by n-gao** [comment]
@sebastianecheverrir could you run the pre-commit hooks? (it seems that files are not formatted correctly)

Then this should be good to land! :)

**Comment by sebastianecheverrir** [comment]
The changes from the pre-commit hooks are now included

**Review by n-gao — APPROVED** [review]
LGTM
